### DISCUSSION TITLE
Add a way to suppress `finally` block with `using`

### DISCUSSION THREAD

**Discussion by MichalPetryka** [dotnet/csharplang#7249] [Language Ideas]
Currently, every `using` on a type with a `Dispose` method causes Roslyn to emit a `try-finally` in the produced IL to guarantee the `Dispose` being called. This can be both a performance and a safety issue.

On the performance side, an exception handling block causes the JIT to block many optimizations of the code, for example inlining. This makes `using` problematic in high performance code that relies on it. Skipping `Dispose` after an exception won't be an issue in most cases since types will usually have finalizers that'll take care of the memory instead and in other cases the dispose might not always be necessary.

With safety, sometimes disposing after an exception poses a risk of use-after-free or process corruption. An example of this is the BCL `ArrayPool<T>`, if you return a rented array while something still holds a reference to it, it can then write to it while something else has re-rented it, corrupting the process state. In other cases you can experience crashes due to native code still accessing the freed memory. Leaking memory is a safer solution then, as @GrabYourPitchforks said a few times: "And if something goes horribly wrong and the consumer doesn't know what to do, we'd generally prefer memory leaks over use-after-free. At least memory leaks will generally trigger watchdog processes to reboot the service.".

For the implementation side, I'd say that an attribute `[SkipUsingFinally]` would be the simplest way to implement this, it'd then be placed either on the `Dispose` or on the type.

One issue I can think of is with inheritance, whether child types should inherit parent `SkipUsingFinally` is something I am not sure of.
I am also not sure whether in case of multiple `return`s inside of the `using` block there should be multiple calls to `Dispose`, before every return or if the compiler should emit jumps to a shared `Dispose` + `return` instead.

Removing this from a type would have to be considered a binary breaking change, since assemblies compiled using older versions of the type would still be missing the `finally` then.

**Comment by CyrusNajmabadi** [comment]
> On the performance side, an exception handling block causes the JIT to block many optimizations of the code, for example inlining. 

I'm wary of justifications like this. It's a point in time aspect about how the runtime works today. But it may not hold in the future. We've seen many runtime perf issues fade and become non-issues over the years. So designing an escape hatch that might not matter in a few years worries me. 

The case of wanting to cleanup always, except in exception conditions, seems super weird to me. This is especially true as try/finally/dispose exist so that non-obvious control flow (like exceptions), won't cause problems. 

If you don't care about the exception case, I think this is better handled with an attribute (which give already stated you would use) and an *analyzer*. 

The analyzer would be simple, and would tell you if there were non-exception exit points reachable that did not Dispose the value. We this you'd have the perf you want, with the safety as well. 

You'd be more verbose. But I think that's a good thing here. I think it would actively be bad for someone to see a `using` and then have to know it didn't run in the cases of exceptions.

**Reply by MichalPetryka** [reply]
The performance side of this is less important here, although AFAIR there are no runtime changes planned that will remove the drawbacks from finally.

Like I've mentioned, `ArrayPool` is an example of a type that should only be freed in case of successful return, since keeping a reference to the array due to an exception can corrupt the state instead of just letting the GC collect the array.

Using an analyzer here wouldn't be really feasible with the attribute being placed on the type since you then wouldn't know which methods are supposed to dispose the object in all cases and which are not. You could then at best block using `using` on it. And placing the attribute on the method using the type shifts the responsibility to the end user of the code, not the library author, making this useless since end user can't guarantee such behaviour to be safe.

As such, you'd need to manually document it to be safe and have the users make sure all the possible places the code can return from have a `Dispose` present, which both clutters the code and can be easily missed when adding a new one.

Since this attribute would be placed on the target type if it'd be a language feature, the users wouldn't need to know that it doesn't run in case of exceptions since the library author guarantees that behaviour to be a correct one.

**Reply by MichalPetryka** [reply]
An example case of a simplification this'd allow:
```cs
byte[] array = ArrayPool<byte>.Shared.Rent(1024);

if (condition_a)
{
    var result = CallA(array);
    ArrayPool<byte>.Shared.Return(array);
    return result;
}

if (condition_b)
{
    var result = CallB(array);
    ArrayPool<byte>.Shared.Return(array);
    return result;
}

if (condition_c)
{
    var result = CallC(array);
    ArrayPool<byte>.Shared.Return(array);
    return result;
}

throw new UnreachableException();
```
could instead be:
```cs
[SkipUsingFinally]
readonly ref struct PooledMemory<T>
{
    private readonly T[] array;
    public PooledMemory(int count) => array = ArrayPool<byte>.Shared.Rent(count);
    public static operator implicit T[](PooledMemory<T> memory) => array;
    public void Dispose() => ArrayPool<byte>.Shared.Return(array);
}

using PooledMemory<byte> array = new(1024);

if (condition_a)
    return CallA(array);

if (condition_b)
    return CallB(array);

if (condition_c)
    return CallC(array);

throw new UnreachableException();
```

**Reply by CyrusNajmabadi** [reply]
You say it's the correct behavior, and the user won't need to know, but I don't think that's possible.  For example, as a user I may want to clean up N things. So I make a wrapper type that itself just wraps an *IDisposable*.  Once I wrap your type as that, we lose the special codegen and we're back to corrupting state. So you must actually block this somehow (which means an analyzer, which means you're back to what I was referring to).

> Like I've mentioned, ArrayPool is an example of a type that should only be freed in case of successful return, since keeping a reference to the array due to an exception can corrupt the state instead of just letting the GC collect the array.

My approach would work for that type.

> Using an analyzer here wouldn't be really feasible with the attribute being placed on the type since you then wouldn't know which methods are supposed to dispose the object in all cases and which are not. 


I don't understand. If the analyzer wouldn't know, then how could the compiler know?  And if there is information to inform the compiler, then it could inform an analyzer. 

> making this useless since end user can't guarantee such behaviour to be safe.

One of two things is true. Either there is enough information to make the determination, in which case an analyser could error if you did this wrong. Or there isn't enough information, and the compiler can't make this safe either. 

> Since this attribute would be placed on the target type if it'd be a language feature, the users wouldn't need to know that it doesn't run in case of exceptions since the library author guarantees that behaviour to be a correct one.

How can it guarantee it?  First, compiling with any other compiler (like vb, f#, etc) would not work. Second, the moment you call through IDisposable, this breaks. This fundamentally can no longer be an IDisposable impl because the information would not carry through (without runtime introspection of the actual instance).

**Reply by CyrusNajmabadi** [reply]
Ok. It sounds like you just want to limit this to ref-structs, thus making it so you could never view this as IDisposable.Dispose. 

In the case, this can def be done with an analyzer as I mentioned originally.  The code would be more verbose. But my point is that I think that's a good thing. I think `using` is bad here as this defies the expectations we've set up from 1.0 about the constructs explicitly existing for non explicit control flow handling. 

Instead of this, of we want less verbosity at the use site, I would say you'd need another keyword than `using`, or an augmentation of the using-statement so it's clear which behavior you are getting.

**Reply by CyrusNajmabadi** [reply]
For example (not actually proposing this as the syntax): `using noexcept var array = new PooledMemory(1024);`.

**Reply by CyrusNajmabadi** [reply]
Note: I'm still confused about the justification here. If this is a ref-struct, it's has to live on the stack, and deeper calls can't hold onto this without violating red safety.  So if an exception happens, they will pop out and not have any refs to the array (you're not exposing the raw array right, just an indirection around it?). 

With that pattern, no one ever sees or can hold onto the heap data you want to return to the pool, keeping things safe.  For something like an array, this means only exposing as a Span when passing things around. Indeed, it's unclear to me why you don't just rent the array and pass the *Span* to it to your methods. 

This would be my expectation for how this has to work *even without exceptions*. Because of you expose the heap information, then someone can still hold onto it even after they return normally, so they can corrupt things.

**Reply by CyrusNajmabadi** [reply]
Oh, I see this line now:

`public static operator implicit T[](PooledMemory<T> memory) => array;`

This is bad even in the absence of exceptions. You have no protection for people screwing up and corrupting this data after you return it to the pool. Use Span (or another ref-struct you know they can't reference from the heap).
